### PR TITLE
Generate error.h automatically

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
There is an overview of error codes in a comment in `include/mbedtls/error.h`. This overview has been out of synch for a long time. This PR enhances `scripts/generate_errors.pl` to update `include/mbedtls/error.h` in addition to `library/error.c`.

The script contains quirks to preserve the current format of `error.h`. It could be simplified if we decide to format the information differently.

Backports: #1205 #1206

### CODE REVIEW

File: scripts/generate_errors.pl [hunk 1]
```diff
@@ -46,6 +45,44 @@
 
 $/ = $line_separator;
 
+=comment
```
Comment (mpg) [inline scripts/generate_errors.pl]: I don't get what the purpose of leaving commented block of code in the script is. Can you explain?

File: scripts/generate_errors.pl [hunk 2]
```diff
@@ -97,7 +137,7 @@
     my $found_hl = grep $_ eq $module_name, @high_level_modules;
     if (!$found_ll && !$found_hl)
     {
-        printf("Error: Do not know how to handle: $module_name\n");
+        printf STDERR ("Error: unknown module name: $module_name\n");
```
Comment (mpg) [inline scripts/generate_errors.pl]: Style: why not `warn`?

File: scripts/generate_errors.pl [hunk 3]
```diff
@@ -3,23 +3,22 @@
 # Generate error.c
 #
 # Usage: ./generate_errors.pl or scripts/generate_errors.pl without arguments,
-# or generate_errors.pl include_dir data_dir error_file
+# or generate_errors.pl include_dir data_dir error_c [error_h]
```
Comment (mpg) [inline scripts/generate_errors.pl]: Can you also update the `Generate error.c` line above?

And, while this is unrelated, we should probably also document that this script checks that no error code is duplicated (I was never sure about that).

File: scripts/generate_errors.pl [hunk 4]
```diff
@@ -193,6 +245,110 @@
...
 $error_format =~ s/HIGH_LEVEL_CODE_CHECKS\n/$hl_code_check/g;
 
-open(ERROR_FILE, ">$error_file") or die "Opening destination file '$error_file': $!";
-print ERROR_FILE $error_format;
-close(ERROR_FILE);
+open(ERROR_FILE, ">$error_c") or die "Opening destination file '$error_c': $!";
+print ERROR_FILE $error_format or die "Writing '$error_c': $!";
+close(ERROR_FILE) or die "Closing '$error_c': $!";
+
+sub check_range
```
Comment (mpg) [inline scripts/generate_errors.pl]: Please document the calling convention for this subroutine, especially as the way arguments are handled (`shift` `shift` `pop` `foreach`) is a bit unusual.

File: scripts/generate_errors.pl [hunk 5]
```diff
@@ -193,6 +245,110 @@
...
+        ++$h_high_entries{$name}{nr};
+        push @{$h_high_entries{$name}{codes}}, $value;
+    }
+}
+foreach my $name (keys %h_low_entries)
+{
+    my $entry = $h_low_entries{$name};
+    my @even = sort {$a <=> $b} @{$entry->{even}};
+    my @odd = sort {$a <=> $b} @{$entry->{odd}};
+    check_range($name, @even);
```
Comment (mpg) [inline scripts/generate_errors.pl]: I might be missing something obvious, but I don't get it: you just sorted that array, so how could `check_range()` fail?

File: scripts/generate_errors.pl [hunk 6]
```diff
@@ -193,6 +245,110 @@
...
+my $h_high_text =
+    join('',
+         map {sprintf(" * %-9s %2d  %d%s\n", $_,
+                      $h_high_entries{$_}{id},
+                      $h_high_entries{$_}{nr},
+                      $h_high_entries{$_}{comment})}
+         sort {$h_high_entries{$a}{sort_key} cmp $h_high_entries{$b}{sort_key}}
+         keys %h_high_entries);
+
+open(ERROR_FILE, "+<$error_h") or die "Opening destination file '$error_h': $!";
```
Comment (mpg) [inline scripts/generate_errors.pl]: I quite like the way `error.h` is updated in place, but this is not consistent with the way other files are generated (from a template in `scripts/data_files`. I feel like this discrepancy should be clearly documented in the comments at the beginning of the script. (In the long run, I'd be in favour of resolving the discrepancy by using this method for other generated files as well, but I understand this might go a bit beyond the scope of this PR, so just documenting it in the meantime is fine.)

File: include/mbedtls/error.h [hunk 7]
```diff
@@ -48,38 +48,37 @@
...
- * PBKDF2     1  0x007C-0x007C
  * HMAC_DRBG  4                  0x0003-0x0009
  * CCM        2                  0x000D-0x000F
  *
  * High-level module nr (3 bits - 0x0...-0x7...)
  * Name      ID  Nr of Errors
  * PEM        1  9
  * PKCS12     1  4 (Started from top)
- * X509       2  20
+ * X509       3  20
```
Comment (mpg) [inline include/mbedtls/error.h]: Ouch, this reveals a mistake that has been made when allocating `MBEDTLS_ERR_X509_FATAL_ERROR`: it should clearly be `0x2A00` not `0x3000`. This is very unfortunate is has already been published and fixing it is an ABI change now (I added it to our list).

I think the script should be changed to still output `2` as the ID of X509; I realise this means adding a hack to the script to get around a mistake in `x509.h` but I think that's the best way to go for now. In the future (when the x509 value is fixed) it could be good for the script to warn about this.

File: scripts/generate_errors.pl [hunk 8]
```diff
@@ -193,6 +245,110 @@
...
+foreach my $name (keys %h_high_entries)
+{
+    my $entry = $h_high_entries{$name};
+    my @codes = sort {$a <=> $b} @{$entry->{codes}};
+    check_range($name, @codes);
+    $entry->{id} = $codes[@codes-1] >> 12;
+    $entry->{comment} = (($codes[0] & 0xf80) == 0x080 ? '' :
+                         ($codes[@codes-1] & 0xf80) == 0xf80 ?
+                         ' (Started from top)' :
+                         ' (Started from middle)');
```
Comment (mpg) [inline scripts/generate_errors.pl]: I don't think we have modules starting from middle so far. Is that added on purpose as a planned extension?

File: scripts/generate_errors.pl [hunk 9]
```diff
@@ -108,12 +148,23 @@
...
+            printf STDERR ("Error: invalid low-level error code $error_code ($error_name)\n");
+            exit 1;
+        }
         $code_check = \$ll_code_check;
         $old_define = \$ll_old_define;
         $white_space = '    ';
     }
     else
     {
+        if ($error_number == 0 || $error_number > 0x7fff ||
```
Comment (Patater) [inline scripts/generate_errors.pl]: Do we want to also catch when an error number doesn't have the same high level module ID as other errors within the same module (that have been seen so far)?

Note that we may need to make an exception for errors in error number allocation we've already made (like `MBEDTLS_ERR_X509_FATAL_ERROR` which doesn't share the same high level module ID as other X.509 errors).

**Comment by irwir** [comment]
The currently generated code always compares `use_ret` to every existing error code, and every time there is a call to `_snprintf()`. This could be improved.
For example, by using `switch` statement:
```
    char *pmsg;
. . .
        // High level error codes
        //
        switch( use_ret )
        {
        // BEGIN generated code
#if defined(MBEDTLS_CIPHER_C)
        case -(MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE):
            pmsg = "CIPHER - The selected feature is not available";
            break;
        case -(MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA):
. . .
. . .
            break;
#endif /* MBEDTLS_X509_USE_C || MBEDTLS_X509_CREATE_C */
        // END generated code
        default:
            pmsg = "";
        }
        if( pmsg[0] == '\0' )
            mbedtls_snprintf( buf, buflen, "UNKNOWN ERROR CODE (%04X)", use_ret );
        else
            mbedtls_snprintf( buf, buflen, pmsg);
        if( use_ret == -(MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE) )
            return;
    }

```
_The whole `error.c` module could be posted here if necessary._

Alternatively, there could be an array of error codes, corresponding array of string pointers (or a structure, containing both error codes and string pointers), and a simple loop to scan error code values.

Any of these changes should be fairly straightforward in implementation.

**Comment by gilles-peskine-arm** [comment]
@irwir I hadn't planned to change that part of the code. Yes, it look a bit nicer, but does it matter? Are common compilers unable to optimize this in the same way as a switch statement?

**Comment by irwir** [comment]
> Are common compilers unable to optimize this in the same way as a switch statement?

The code is written to check every condition sequentially.
I am not aware of compilers being smart enough to optimize according to unwritten programmer's intentiions. 
Using `else if` instead of `if` will be somewhat better, but `switch` makes code more clear and uniform.

File: scripts/generate_errors.pl [hunk 10]
```diff
@@ -46,6 +45,44 @@
 
 $/ = $line_separator;
 
+=comment
+open(ERROR_H, "<", $error_h) or die "Opening error.h";
+my %sets;
+my $section == 0;
```
Comment (tom-cosgrove-arm) [inline scripts/generate_errors.pl:51]: this should be `=`

Comment (gilles-peskine-arm) [reply scripts/generate_errors.pl:51]: Or maybe given the size of the patch, if we revive this, we'll rewrite the whole script in python.
